File: Floor.java
public class Floor {

	final int floorNum;
	boolean fireKey;
	boolean fireKeyVisited;
	
	public Floor(int floorNum) {
		this.floorNum = floorNum;
		fireKey = false;
		fireKeyVisited = false;
	}
	
	public void callUp() {
		Main.elev.registers.setFloorCalled(floorNum, 1);
	}

	public void callDown() {
		Main.elev.registers.setFloorCalled(floorNum, -1);
	}
	
	public void insertKey() {
		fireKey = true;
		fireKeyVisited = false;
		this.callDown();
	}

	public void removeKey() {
		fireKey = false;
		fireKeyVisited = false;
		Main.elev.registers.reset();
	}

}

File: Floor.cpp
#include <cstdlib>
#include <iostream>
#include <ctime>
#include <unistd.h>

using namespace std;

class Floor{
    private static Elevator elev;
    
    public static void init(Elevator elevator){
        elev = elevator;
    }
    
    public static void callUp(int currentFloor){
        if (currentFloor == 3)
            return;
        
        RegisterBank.illuminateButton(currentFloor + 4);
        RegisterBank.setFloor(currentFloor);
    }
    
    public static void callDown(int currentFloor){
        if (currentFloor = 0)
            return;
        
        RegisterBank.illuminateButton(currentFloor);
        RegisterBank.setFloor(currentFloor);
    }
    
    public static void callKey(int currentFloor){
        elev.setKey(true);
        RegisterBank.setFloor(currentFloor);
    }
}